### PR TITLE
SHT4x driver for wipperSnapper

### PR DESCRIPTION
**Pull request by TheGodMotherG**
first attempt

### CODE REVIEW

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
+/*!
+ * @file WipperSnapper_I2C_Driver_Sht4x.h
+ *
+ * Device driver for an SHT Humidity and Temperature sensor.
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:4]: Change to: "Device drive for a SHT4x Humidity and Temperature Sensor"

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ *
+ * MIT license, all text here must be included in any redistribution.
+ *
+ */
+
+#ifndef WipperSnapper_I2C_Driver_Sht4x_H
+#define WipperSnapper_I2C_Driver_Sht4x_H
+
+#include "WipperSnapper_I2C_Driver.h"
+#include <Adafruit_Sht4x.h>
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:20]: This library #include should be: "#include "Adafruit_SHT4x.h", see: https://github.com/adafruit/Adafruit_SHT4X/blob/master/examples/SHT4test/SHT4test.ino#L11

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 3]
```diff
@@ -0,0 +1,148 @@
+/*!
+ * @file WipperSnapper_I2C_Driver_Sht4x.h
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:2]: Change file name to `WipperSnapper_I2C_Driver_SHT4x.h`

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+#include "WipperSnapper_I2C_Driver.h"
+#include <Adafruit_Sht4x.h>
+
+/**************************************************************************/
+/*!
+    @brief  Class that provides a sensor driver for the Sht4x temperature
+            and humidity sensor.
+*/
+/**************************************************************************/
+class WipperSnapper_I2C_Driver_Sht4x : public WipperSnapper_I2C_Driver {
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:28]: Change class name to `WipperSnapper_I2C_Driver_SHT4x` across the file.

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  /*******************************************************************************/
+  /*!
+      @brief    Initializes the Sht4x sensor and begins I2C.
+      @returns  True if initialized successfully, False otherwise.
+
+  */
+  /*******************************************************************************/
+  bool begin() {
+    _sht = new Adafruit_Sht4x();
+    bool isInit = _sht->begin(_i2c, (int21_t)_sensorAddress);
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:62]: sht begin() signature only needs the wire interface (see: https://github.com/adafruit/Adafruit_SHT4X/blob/master/Adafruit_SHT4x.h#L118)

Change this line to:
`bool isInit = _sht->begin(_i2c);`

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    return isInit;
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:63]: Looking at the demo file, https://github.com/adafruit/Adafruit_SHT4X/blob/master/examples/SHT4test/SHT4test.ino#L31, you can set the precision.

Add the line:
`_sht4->setPrecision(SHT4X_HIGH_PRECISION);`

Also, you'll want to add another line below this which calls `setHeater()`, https://github.com/adafruit/Adafruit_SHT4X/blob/master/examples/SHT4test/SHT4test.ino#L47

File: src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  void enableSensorRelativeHumidity() {
+    _sht_humidity = _sht->getHumiditySensor();
+  }
+
+  /*******************************************************************************/
+  /*!
+      @brief    Disables the Sht4x's temperature sensor.
+  */
+  /*******************************************************************************/
+  void disableSensorAmbientTemperature() {
```
Comment (brentru) [inline src/components/i2c/drivers/WipperSnapper_I2C_Driver_Sht4x.h:89]: Remove this function, not required.

**Review by brentru — CHANGES_REQUESTED** [review]
Requesting modifications before I test.

You also need to:
* Include the SHT4x driver in the I2C Component
  * See this recent pull request for an example: https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/262/files#diff-78ec1584747f78ea74c379069a73dd165575fd5901c93829b4967d7bdebfad80
* Include handling code for the I2C Component to create the sht4x driver 
  * See this recent pull request for an example: https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/262/files#diff-3dd55e689e75bcb8859a65746742c85d8b634ee427ba635ece3bd5d91711eee6
* Inclue the Adafruit SHT4x library within the library.properties file so it can be installed automatically.
  * See this recent pull request as an example: https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/262/files#diff-f267f9ae2f5d5d56a1dbb305e95d40d4408d9ee7f71357f1f67fb18c29b7f082

**Comment by TheGodMotherG** [comment]
I uploaded all the requested files i named the new one SHT4x instead of sht4x and the other 2 files as well. It wont let me delete the old one but all the new files drivers and libraries are uploaded please take a look

**Comment by TheGodMotherG** [comment]
I did my best to do all the recommended changes by brentru but still failed the test if anyone has any ideas how to help i would greatly appreciate it,

**Comment by danielfrankcom** [comment]
It seems like `Adafruit SHT4x Library` needs to be added to the `depends` property in `library.properties`.

The build is failing with an error message that seems to indicate a missing dependency.

```
  /home/runner/Arduino/libraries/Adafruit_Wippersnapper_Arduino/src/components/i2c/drivers/WipperSnapper_I2C_Driver_SHT4x.h:20:10: fatal error: Adafruit_SHT4x.h: No such file or directory
   #include "Adafruit_SHT4x.h"
            ^~~~~~~~~~~~~~~~~~
  compilation terminated.
```

There is a list of installed libraries directly above this error message in the logs, and it doesn't include the required library that is being imported. 

`<repo root>/library.properties` contains a `depends` field with a list of dependencies. I suspect this is what the build server uses to install the required packages before running the build.

According to the Arduino IDE, there is a package named with the exact text "Adafruit SHT4x Library", so if that's added to the list of dependencies hopefully the build will pass.

image

**Comment by brentru** [comment]
Closing this PR in favor of https://github.com/adafruit/Adafruit_Wippersnapper_Arduino/pull/296 which provides tested support for SHT4x. Will be released in next version.

@TheGodMotherG  Thank you for adding the sensor definition JSON and your work here.

**Pull request closed without merging**